typing/typecore.ml

```diff
@@ -2832,23 +2830,42 @@ let collect_apply_args env funct ignore_labels ty_fun ty_fun0 sargs =
                 if not optional && is_optional l' then
                   Location.prerr_warning sarg.pexp_loc
                     (Warnings.Nonoptional_label (Asttypes.string_of_label l));
-                remaining_sargs, use_arg sarg l'
+                remaining_sargs, Some (sarg, l')
             | None ->
-                sargs,
-                if optional && List.mem_assoc Nolabel sargs then
-                  eliminate_optional_arg ()
-                else begin
-                  (* No argument was given for this parameter, we abstract over
-                     it. *)
-                  may_warn funct.exp_loc
-                    (Warnings.Non_principal_labels "commuted an argument");
-                  Omitted { ty_arg; level = lv }
-                end
+                sargs, None
         in
-        loop ty_ret ty_ret0 ((l, arg) :: rev_args) remaining_sargs
-    | _ ->
-      (* We're not looking at a *known* function type anymore. *)
-      collect_unknown_apply_args env funct ty_fun0 rev_args sargs
+        match get_desc ty_fun', get_desc (expand_head env ty_fun0) with
+        | Tarrow (_, ty_arg, ty_ret, _), Tarrow (_, ty_arg0, ty_ret0, _) ->
+            let use_arg sarg l' =
+              let wrapped_in_some = optional && not (is_optional l') in
+              if wrapped_in_some then
+                may_warn sarg.pexp_loc
+                  (not_principal "using an optional argument here");
+              Arg (Known_arg { sarg; ty_arg; ty_arg0; wrapped_in_some })
+            in
+            let eliminate_optional_arg () =
+              may_warn funct.exp_loc
+                (Warnings.Non_principal_labels "eliminated optional argument");
+              Arg (Eliminated_optional_arg { ty_arg; level = lv })
+            in
+            let arg =
+              match arg_opt with
+              | Some (sarg, l') -> use_arg sarg l'
+              | None ->
+                  if optional && List.mem_assoc Nolabel sargs then
+                    eliminate_optional_arg ()
+                  else begin
+                    (* No argument was given for this parameter, we abstract
+                      over it. *)
+                    may_warn funct.exp_loc
+                      (Warnings.Non_principal_labels "commuted an argument");
+                    Omitted { ty_arg; level = lv }
+                  end
+            in
+            loop ty_ret ty_ret0 ((l, arg) :: rev_args) remaining_sargs
+        | _ ->
+          assert false
```

Comment:
Reading the code, I feel that the refactoring is missing some comments to explain why the change in structure is worth the introduction of this new `assert false`.

Reply:
I would go further and suggest to refactor this further to get rid of the new `assert false`, it doesn't look too difficult.

Reply:
Yep, the new commit is reasonable and corresponds to what I had in mind; I approve it. I haven't looked at the others, though.

Reply:
The other commits look good to me, partially because I like extracting auxiliary functions outside of complex pattern matches.